[MediaSource] canplay event does not fire for video elements with renderers when first sample has a pts greater than 0

aestes opened the pull request:
#### d0b6e889f00ee859d4f9ed8f7119a3ed3f5ad7cf
<pre>
[MediaSource] canplay event does not fire for video elements with renderers when first sample has a pts greater than 0
<a href="https://bugs.webkit.org/show_bug.cgi?id=263415">https://bugs.webkit.org/show_bug.cgi?id=263415</a>
rdar://105169372

Reviewed by Jer Noble.

Some MediaSource content adds a source buffer where the first decoded sample has a presentation
start time that is slightly greater than 0. This exposes a bug in WebCoreDecompressionSession where
the &quot;canplay&quot; event is never dispatched because a sample whose pts matches the playback
synchronizer&apos;s current time is never decoded. This bug is tracked by webkit.org/b/263413.

Instead of addressing the issue in WebCoreDecompressionSession, this commit addresses the reason
why WebCoreDecompressionSession was being used to begin with for video elements with renderers.
In 252183@main, MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer() was modified to return
false when a video element&apos;s bounds were empty to address a failure in
fast/canvas/webgl/texImage2D-mse-flipY-false.html, but this turned out to be overly restrictive.
-[AVSampleBufferDisplayLayer copyDisplayedPixelBuffer] actually does work properly when its bounds
are empty; it merely fails when the video element does not have a renderer. 266162@main later added
a renderer check to MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer(), making the
empty-bounds checks unnecessary.

This commit removes the empty-bounds checks from
MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer(), ensuring we use an
AVSampleBufferDisplayLayer for decoding (which is not tripped up by the pts issue described above)
in all situations where the video element has a renderer.

Added tests to verify the following behaviors for video elements with renderers:
- video-in-canvas rendering continues to work via AVSampleBufferDisplayLayer with zero-sized video elements
- &quot;canplay&quot; event is dispatched for zero-sized video elements

Also added a failing test for the issue tracked in webkit.org/b/263413 and marked it as such in
TestExpectations.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-false-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-false.html: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-true-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-true.html: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-expected.txt: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-without-renderer-expected.txt: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-without-renderer.html: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay.html: Added.
* LayoutTests/media/media-source/resources/first-sample-pts-non-zero/LICENSE: Added.
* LayoutTests/media/media-source/resources/first-sample-pts-non-zero/first-sample-pts-non-zero.mp4: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):

Canonical link: <a href="https://commits.webkit.org/269689@main">https://commits.webkit.org/269689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74a855f08ed4934791fe8027b9e130dd1ff7f1e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23257 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24384 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/25526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23526 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23818 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/25526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23514 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/930 "Found 1 new test failure: fast/scrolling/ios/hit-testing-iframe-006.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20146 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26029 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/744 "Found 9 new test failures: imported/w3c/web-platform-tests/css/css-text/i18n/ja/css-text-line-break-ja-pr-normal.html, imported/w3c/web-platform-tests/css/css-text/i18n/ja/css-text-line-break-ja-pr-strict.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-004.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-005.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-006.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-overflow-001.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-wbr-nobr-001.html, imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-wbr-nobr-002.html, imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27190 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21203 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21291 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18498 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/692 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20811 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5553 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/993 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on LayoutTests/media/media-source/media-source-first-sample-pts-non-zero.html:
```diff
@@ -0,0 +1,42 @@
+    var video = document.createElement("video");
+    video.classList.add("zeroSize");
+    document.body.appendChild(video);
+    video.addEventListener("canplay", event => {
```
a bit more modern JS would be to make this async and do:
```
await waitFor(video, 'canplay');
testRunner.notifyDone();
```
(make this into a onload=test())

jyavenard commented on LayoutTests/media/media-source/media-source-first-sample-pts-non-zero.html:
```diff
@@ -0,0 +1,42 @@
+        testRunner.waitUntilDone();
+
+    var source = new MediaSource();
+    source.addEventListener("sourceopen", async event => {
```
```
await waitFor(source, 'sourceopen');
var buffer = source.addSourceBuffer("video/mp4");
...
```

jyavenard commented on LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-without-renderer-expected.txt:
```diff
@@ -0,0 +1,3 @@
+CONSOLE MESSAGE: This test verifies that the "canplay" event fires on a video element with a MediaSource that loads a video track whose first sample has a presentation timestamp greater than 0 when the video element is not in the DOM.
```
the issue here is that per spec, we should stall that video.
That we allow a leeway near 0 is entirely user-agent dependent.

aestes replied:
That's fair, although Chrome and Firefox both dispatch "canplay" on this test case so we should too for compatibility. I'll look more closely at what the exact leeway we should implement in webkit.org/b/263413.

aestes committed:
[MediaSource] canplay event does not fire for video elements with renderers when first sample has a pts greater than 0
https://bugs.webkit.org/show_bug.cgi?id=263415
rdar://105169372

Reviewed by Jer Noble.

Some MediaSource content adds a source buffer where the first decoded sample has a presentation
start time that is slightly greater than 0. This exposes a bug in WebCoreDecompressionSession where
the "canplay" event is never dispatched because a sample whose pts matches the playback
synchronizer's current time is never decoded. This bug is tracked by webkit.org/b/263413.

Instead of addressing the issue in WebCoreDecompressionSession, this commit addresses the reason
why WebCoreDecompressionSession was being used to begin with for video elements with renderers.
In 252183@main, MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer() was modified to return
false when a video element's bounds were empty to address a failure in
fast/canvas/webgl/texImage2D-mse-flipY-false.html, but this turned out to be overly restrictive.
-[AVSampleBufferDisplayLayer copyDisplayedPixelBuffer] actually does work properly when its bounds
are empty; it merely fails when the video element does not have a renderer. 266162@main later added
a renderer check to MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer(), making the
empty-bounds checks unnecessary.

This commit removes the empty-bounds checks from
MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer(), ensuring we use an
AVSampleBufferDisplayLayer for decoding (which is not tripped up by the pts issue described above)
in all situations where the video element has a renderer.

Added tests to verify the following behaviors for video elements with renderers:
- video-in-canvas rendering continues to work via AVSampleBufferDisplayLayer with zero-sized video elements
- "canplay" event is dispatched for zero-sized video elements

Also added a failing test for the issue tracked in webkit.org/b/263413 and marked it as such in
TestExpectations.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-false-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-false.html: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-true-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/texImage2D-mse-in-dom-flipY-true.html: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-expected.txt: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-without-renderer-expected.txt: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay-without-renderer.html: Added.
* LayoutTests/media/media-source/media-source-first-sample-pts-non-zero-canplay.html: Added.
* LayoutTests/media/media-source/resources/first-sample-pts-non-zero/LICENSE: Added.
* LayoutTests/media/media-source/resources/first-sample-pts-non-zero/first-sample-pts-non-zero.mp4: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::shouldEnsureLayer const):

Canonical link: https://commits.webkit.org/269689@main
